### DISCUSSION TITLE
Defining a method decorator which uses the self attribute

### DISCUSSION THREAD

**Discussion by Lotram** [microsoft/pyright#8831] [General]
Hi, my goal is quite simple: to create a method decorator which uses the `self` attribute from the original method.

Here is what wrote:
```
from functools import wraps
from typing import Callable, Concatenate, ParamSpec, TypeVar


class MyClass:
    def __init__(self, counter: int = 0) -> None:
        self.counter = counter

    def hello(self, name: str) -> str:
        self.counter += 1
        return f"Hello, {name}!"


T = TypeVar("T", bound=MyClass)
P = ParamSpec("P")
R = TypeVar("R", covariant=True)


def decorate(
    hello: Callable[Concatenate[T, P], R]
) -> Callable[Concatenate[T, P], R]:
    @wraps(hello)
    def decorated(self: T, /, *args: P.args, **kwargs: P.kwargs) -> R:
        print("counter:", self.counter)
        return hello(self, *args, **kwargs)

    return decorated



MyClass.hello = decorate(MyClass.hello)
```

using pyright 1.1.377, I get the following error:
```
  test.py:28:17 - error: Cannot assign to attribute "hello" for class "type[MyClass]"
    Type "(MyClass, name: str) -> str" is incompatible with type "(self: MyClass, name: str) -> str"
      Missing keyword parameter "self"
        Position-only parameter mismatch; parameter "self" is not position-only
        Position-only parameter mismatch; expected 1 but received 0 (reportAttributeAccessIssue)
1 error, 0 warnings, 0 informations 
```
I've read several threads about typing positional-only parameters  such as [this one](https://discuss.python.org/t/ambiguities-about-positional-only-parameters/42328) and [that one](https://discuss.python.org/t/allow-self-binding-for-generic-paramspec/50948) but couldn't find an answer

Could you help me find the proper fix, please ?

**Comment by erictraut** [comment]
You can fix this by explicitly making `self` a positional-only parameter in the `hello` method:

```python
class MyClass:
    ...
    def hello(self, /, name: str) -> str: ...
```

There has been some discussion in typing circles of making all `self` and `cls` parameters implicitly positional-only. There seems to be general agreement that this would be a good thing, but the typing spec hasn't been officially updated as such. If and when the spec is updated, I will change pyright's behavior accordingly.

**Comment by Lotram** [comment]
Thank you ! I was focusing on fixing my decorator, and forgot to look at the actual method
